src/uu/pmap/src/smaps_format_parser.rs

```diff
@@ -0,0 +1,556 @@
+// This file is part of the uutils procps package.
+//
+// For the full copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+use crate::maps_format_parser::{parse_map_line, MapLine};
+use std::io::{Error, ErrorKind};
+
+// Represents a parsed single entry from /proc/<PID>/smaps for the extended formats.
+#[derive(Debug, Clone, Default, PartialEq)]
+pub struct SmapEntry {
+    pub map_line: MapLine,
+    pub kernel_page_size_in_kb: u64,
+    pub mmu_page_size_in_kb: u64,
+    pub rss_in_kb: u64,
+    pub pss_in_kb: u64,
+    pub pss_dirty_in_kb: u64,
+    pub shared_clean_in_kb: u64,
+    pub shared_dirty_in_kb: u64,
+    pub private_clean_in_kb: u64,
+    pub private_dirty_in_kb: u64,
+    pub referenced_in_kb: u64,
+    pub anonymous_in_kb: u64,
+    pub ksm_in_kb: u64,
+    pub lazy_free_in_kb: u64,
+    pub anon_huge_pages_in_kb: u64,
+    pub shmem_pmd_mapped_in_kb: u64,
+    pub file_pmd_mapped_in_kb: u64,
+    pub shared_hugetlb_in_kb: u64,
+    pub private_hugetlb_in_kb: u64,
+    pub swap_in_kb: u64,
+    pub swap_pss_in_kb: u64,
+    pub locked_in_kb: u64,
+    pub thp_eligible: u64,
+    pub vmflags: String,
+}
+
+// Parses entries from /proc/<PID>/smaps. See
+// https://www.kernel.org/doc/html/latest/filesystems/proc.html for details about the expected
+// format.
+//
+// # Errors
+//
+// Will return an `Error` if the format is incorrect.
+pub fn parse_smap_entries(contents: &str) -> Result<Vec<SmapEntry>, Error> {
+    let mut smap_entries = Vec::new();
+    let mut smap_entry = SmapEntry::default();
+
+    for (i, line) in contents.lines().enumerate() {
+        let map_line = parse_map_line(line);
+        if let Ok(map_line) = map_line {
+            if i > 0 {
+                smap_entries.push(smap_entry.clone());
+                smap_entry = SmapEntry::default();
+            }
+            smap_entry.map_line = map_line;
+        } else {
+            let (key, val) = line
+                .split_once(':')
+                .ok_or_else(|| Error::from(ErrorKind::InvalidData))?;
+            let val = val.trim();
+
+            if key == "VmFlags" {
+                smap_entry.vmflags = val.into();
+            } else {
+                let val = if let Some(val) = val.strip_suffix(" kB") {
+                    get_smap_item_value(val)?
+                } else {
+                    get_smap_item_value(val)?
+                };
```

Comment:
While this works fine, it looks unusual with the `if` and `else` blocks being identical. I think you can improve it with something like:
```suggestion
                let val = val.strip_suffix(" kB").unwrap_or(val);
                let val = get_smap_item_value(val)?;
```